reanalysis/hail_filter_and_classify.py

```diff
@@ -0,0 +1,676 @@
+    - CADD & REVEL above threshold (switched to consensus), or
+    - Massive cross-tool consensus
+    - polyphen and sift are evaluated per-consequence
+    :param matrix:
+    :param config:
+    :return:
+    """
+
+    return matrix.annotate_rows(
+        info=matrix.info.annotate(
+            Class4=hl.if_else(
+                (
+                    (matrix.info.cadd > config.get('cadd'))
+                    & (matrix.info.revel > config.get('revel'))
+                )
+                | (
+                    (
+                        matrix.vep.transcript_consequences.any(
+                            lambda x: hl.or_else(x.sift_score, MISSING_FLOAT_HI)
+                            < config.get('sift')
+                        )
+                    )
+                    & (
+                        matrix.vep.transcript_consequences.any(
+                            lambda x: hl.or_else(x.polyphen_score, MISSING_FLOAT_LO)
+                            > config.get('polyphen')
+                        )
+                    )
+                    & (
+                        (matrix.info.mutationtaster.contains('D'))
+                        | (matrix.info.mutationtaster == 'missing')
+                    )
+                    & (matrix.info.gerp_rs >= config.get('gerp'))
+                    & (matrix.info.eigen_phred > config.get('eigen'))
+                ),
+                ONE_INT,
+                MISSING_INT,
+            )
+        )
+    )
+
+
+def filter_matrix_by_ac(
+    matrix_data: hl.MatrixTable, config: Dict[str, Any]
+) -> hl.MatrixTable:
+    """
+
+    :param matrix_data:
+    :param config:
+    :return:
+    """
+
+    # count the samples in the VCF, and use to decide whether to implement
+    # 'common within this joint call' as a filter
+    # if we reach the sample threshold, filter on AC
+    if matrix_data.count_cols() >= config['min_samples_to_ac_filter']:
+        matrix_data = matrix_data.filter_rows(
+            matrix_data.info.AC <= matrix_data.info.AN // config['ac_filter_percentage']
+        )
+    return matrix_data
+
+
+def filter_matrix_by_variant_attributes(matrix_data: hl.MatrixTable) -> hl.MatrixTable:
+    """
+    filter MT to rows with normalised, high quality variants
+    :param matrix_data:
+    :return:
+    """
+
+    # hard filter for quality; assuming data is well normalised in pipeline
+    matrix_data = matrix_data.filter_rows(
+        (
+            matrix_data.filters.length() == 0
+        )  # clarify with FILTER='PASS' & GATK SNP tranches
+        & (hl.len(matrix_data.alleles) == 2)
+        & (matrix_data.alleles[1] != '*')
+    )
+    return matrix_data
+
+
+def annotate_using_vep(matrix_data: hl.MatrixTable) -> hl.MatrixTable:
+    """
+    runs VEP annotation on the MT
+    :param matrix_data:
+    :return:
+    """
+
+    # throw in a repartition here (annotate even chunks in parallel)
+    matrix_data = matrix_data.repartition(150, shuffle=False)
```

Comment:
Should the number of partitions depend on the input size instead?

Reply:
I'm having a look at this now. 

Do you know offhand whether a repartition with a `shuffle` would affect row ordering, or is it strictly to do with the data representation behind the scenes?

Reply:
looks like a full shuffle then a re-sort: https://discuss.hail.is/t/shuffling-and-writing-a-matrixtable-appears-to-run-the-shuffle-op-twice/2202/2

Reply:
With `shuffle=False`, you can only combine existing partitions, i.e. you can never introduce more partitions and the resulting partitions might not all be equally sized. `shuffle=True` requires distributing the data across nodes to create even partitions. On Dataproc / Spark, in practice this means you'll have to use non-preemptible nodes, as otherwise any preemption will require the whole shuffle to restart. This shouldn't be as much of a problem with the query service anymore, as there will be dedicated non-preemptible nodes for shuffling.

Reply:
The row ordering is still only given by the row key, i.e. this isn't really related to shuffling itself.